File: src/client/acontext-ts/src/resources/spaces.ts
/**
 * Spaces endpoints.
 */

import { RequesterProtocol } from '../client-types';
import { buildParams } from '../utils';
import {
  ListSpacesOutput,
  ListSpacesOutputSchema,
  SearchResultBlockItem,
  SearchResultBlockItemSchema,
  Space,
  SpaceSchema,
  SpaceSearchResult,
  SpaceSearchResultSchema,
} from '../types';

export class SpacesAPI {
  constructor(private requester: RequesterProtocol) { }

  async list(options?: {
    limit?: number | null;
    cursor?: string | null;
    timeDesc?: boolean | null;
  }): Promise<ListSpacesOutput> {
    const params = buildParams({
      limit: options?.limit ?? null,
      cursor: options?.cursor ?? null,
      time_desc: options?.timeDesc ?? null,
    });
    const data = await this.requester.request('GET', '/space', {
      params: Object.keys(params).length > 0 ? params : undefined,
    });
    return ListSpacesOutputSchema.parse(data);
  }

  async create(options?: {
    configs?: Record<string, unknown>;
  }): Promise<Space> {
    const payload: Record<string, unknown> = {};
    if (options?.configs !== undefined) {
      payload.configs = options.configs;
    }
    const data = await this.requester.request('POST', '/space', {
      jsonData: Object.keys(payload).length > 0 ? payload : undefined,
    });
    return SpaceSchema.parse(data);
  }

  async delete(spaceId: string): Promise<void> {
    await this.requester.request('DELETE', `/space/${spaceId}`);
  }

  async updateConfigs(
    spaceId: string,
    options: {
      configs: Record<string, unknown>;
    }
  ): Promise<void> {
    const payload = { configs: options.configs };
    await this.requester.request('PUT', `/space/${spaceId}/configs`, {
      jsonData: payload,
    });
  }

  async getConfigs(spaceId: string): Promise<Space> {
    const data = await this.requester.request('GET', `/space/${spaceId}/configs`);
    return SpaceSchema.parse(data);
  }

  /**
   * Perform experience search within a space.
   * 
   * This is the most advanced search option that can operate in two modes:
   * - fast: Quick semantic search (default)
   * - agentic: Iterative search with AI-powered refinement
   * 
   * @param spaceId - The UUID of the space
   * @param options - Search options
   * @returns SpaceSearchResult containing cited blocks and optional final answer
   */
  async experienceSearch(
    spaceId: string,
    options: {
      query: string;
      limit?: number | null;
      mode?: 'fast' | 'agentic' | null;
      semanticThreshold?: number | null;
      maxIterations?: number | null;
    }
  ): Promise<SpaceSearchResult> {
    const params = buildParams({
      query: options.query,
      limit: options.limit ?? null,
      mode: options.mode ?? null,
      semantic_threshold: options.semanticThreshold ?? null,
      max_iterations: options.maxIterations ?? null,
    });
    const data = await this.requester.request(
      'GET',
      `/space/${spaceId}/experience_search`,
      { params: Object.keys(params).length > 0 ? params : undefined }
    );
    return SpaceSearchResultSchema.parse(data);
  }

  /**
   * Perform semantic glob (glob) search for page/folder titles.
   * 
   * Searches specifically for page/folder titles using semantic similarity,
   * similar to a semantic version of the glob command.
   * 
   * @param spaceId - The UUID of the space
   * @param options - Search options
   * @returns List of SearchResultBlockItem objects matching the query
   */
  async semanticGlobal(
    spaceId: string,
    options: {
      query: string;
      limit?: number | null;
      threshold?: number | null;
    }
  ): Promise<SearchResultBlockItem[]> {
    const params = buildParams({
      query: options.query,
      limit: options.limit ?? null,
      threshold: options.threshold ?? null,
    });
    const data = await this.requester.request(
      'GET',
      `/space/${spaceId}/semantic_glob`,
      { params: Object.keys(params).length > 0 ? params : undefined }
    );
    return (data as unknown[]).map((item) =>
      SearchResultBlockItemSchema.parse(item)
    );
  }

  /**
   * Perform semantic grep search for content blocks.
   * 
   * Searches through content blocks (actual text content) using semantic similarity,
   * similar to a semantic version of the grep command.
   * 
   * @param spaceId - The UUID of the space
   * @param options - Search options
   * @returns List of SearchResultBlockItem objects matching the query
   */
  async semanticGrep(
    spaceId: string,
    options: {
      query: string;
      limit?: number | null;
      threshold?: number | null;
    }
  ): Promise<SearchResultBlockItem[]> {
    const params = buildParams({
      query: options.query,
      limit: options.limit ?? null,
      threshold: options.threshold ?? null,
    });
    const data = await this.requester.request(
      'GET',
      `/space/${spaceId}/semantic_grep`,
      { params: Object.keys(params).length > 0 ? params : undefined }
    );
    return (data as unknown[]).map((item) =>
      SearchResultBlockItemSchema.parse(item)
    );
  }
}

File: src/client/acontext-py/src/acontext/resources/spaces.py
"""
Spaces endpoints.
"""

from collections.abc import Mapping
from typing import Any, List

from .._utils import build_params
from ..client_types import RequesterProtocol
from ..types.space import (
    ListSpacesOutput,
    SearchResultBlockItem,
    Space,
    SpaceSearchResult,
)


class SpacesAPI:
    def __init__(self, requester: RequesterProtocol) -> None:
        self._requester = requester

    def list(
        self,
        *,
        limit: int | None = None,
        cursor: str | None = None,
        time_desc: bool | None = None,
    ) -> ListSpacesOutput:
        """List all spaces in the project.

        Args:
            limit: Maximum number of spaces to return. Defaults to None.
            cursor: Cursor for pagination. Defaults to None.
            time_desc: Order by created_at descending if True, ascending if False. Defaults to None.

        Returns:
            ListSpacesOutput containing the list of spaces and pagination information.
        """
        params = build_params(limit=limit, cursor=cursor, time_desc=time_desc)
        data = self._requester.request("GET", "/space", params=params or None)
        return ListSpacesOutput.model_validate(data)

    def create(self, *, configs: Mapping[str, Any] | None = None) -> Space:
        """Create a new space.

        Args:
            configs: Optional space configuration dictionary. Defaults to None.

        Returns:
            The created Space object.
        """
        payload: dict[str, Any] = {}
        if configs is not None:
            payload["configs"] = configs
        data = self._requester.request("POST", "/space", json_data=payload)
        return Space.model_validate(data)

    def delete(self, space_id: str) -> None:
        """Delete a space by its ID.

        Args:
            space_id: The UUID of the space to delete.
        """
        self._requester.request("DELETE", f"/space/{space_id}")

    def update_configs(
        self,
        space_id: str,
        *,
        configs: Mapping[str, Any],
    ) -> None:
        """Update space configurations.

        Args:
            space_id: The UUID of the space.
            configs: Space configuration dictionary.
        """
        payload = {"configs": configs}
        self._requester.request("PUT", f"/space/{space_id}/configs", json_data=payload)

    def get_configs(self, space_id: str) -> Space:
        """Get space configurations.

        Args:
            space_id: The UUID of the space.

        Returns:
            Space object containing the configurations.
        """
        data = self._requester.request("GET", f"/space/{space_id}/configs")
        return Space.model_validate(data)

    def experience_search(
        self,
        space_id: str,
        *,
        query: str,
        limit: int | None = None,
        mode: str | None = None,
        semantic_threshold: float | None = None,
        max_iterations: int | None = None,
    ) -> SpaceSearchResult:
        """Perform experience search within a space.

        This is the most advanced search option that can operate in two modes:
        - fast: Quick semantic search (default)
        - agentic: Iterative search with AI-powered refinement

        Args:
            space_id: The UUID of the space.
            query: The search query string.
            limit: Maximum number of results to return (1-50, default 10).
            mode: Search mode, either "fast" or "agentic" (default "fast").
            semantic_threshold: Cosine distance threshold (0=identical, 2=opposite).
            max_iterations: Maximum iterations for agentic search (1-100, default 16).

        Returns:
            SpaceSearchResult containing cited blocks and optional final answer.
        """
        params = build_params(
            query=query,
            limit=limit,
            mode=mode,
            semantic_threshold=semantic_threshold,
            max_iterations=max_iterations,
        )
        data = self._requester.request(
            "GET", f"/space/{space_id}/experience_search", params=params or None
        )
        return SpaceSearchResult.model_validate(data)

    def semantic_glob(
        self,
        space_id: str,
        *,
        query: str,
        limit: int | None = None,
        threshold: float | None = None,
    ) -> List[SearchResultBlockItem]:
        """Perform semantic glob (glob) search for page/folder titles.

        Searches specifically for page/folder titles using semantic similarity,
        similar to a semantic version of the glob command.

        Args:
            space_id: The UUID of the space.
            query: Search query for page/folder titles.
            limit: Maximum number of results to return (1-50, default 10).
            threshold: Cosine distance threshold (0=identical, 2=opposite).

        Returns:
            List of SearchResultBlockItem objects matching the query.
        """
        params = build_params(query=query, limit=limit, threshold=threshold)
        data = self._requester.request(
            "GET", f"/space/{space_id}/semantic_glob", params=params or None
        )
        return [SearchResultBlockItem.model_validate(item) for item in data]

    def semantic_grep(
        self,
        space_id: str,
        *,
        query: str,
        limit: int | None = None,
        threshold: float | None = None,
    ) -> List[SearchResultBlockItem]:
        """Perform semantic grep search for content blocks.

        Searches through content blocks (actual text content) using semantic similarity,
        similar to a semantic version of the grep command.

        Args:
            space_id: The UUID of the space.
            query: Search query for content blocks.
            limit: Maximum number of results to return (1-50, default 10).
            threshold: Cosine distance threshold (0=identical, 2=opposite).

        Returns:
            List of SearchResultBlockItem objects matching the query.
        """
        params = build_params(query=query, limit=limit, threshold=threshold)
        data = self._requester.request(
            "GET", f"/space/{space_id}/semantic_grep", params=params or None
        )
        return [SearchResultBlockItem.model_validate(item) for item in data]
